src/inference/dreamEubo.ad.js

```diff
@@ -0,0 +1,189 @@
+// Estimates the gradient of EUBO objective based on a list of records generated
+// from dreamSample, each of them includes the trace and the observation values.
+
+'use strict';
+'use ad';
+
+var _ = require('underscore');
+var assert = require('assert');
+var util = require('../util');
+var ad = require('../ad');
+var paramStruct = require('../paramStruct');
+var Trace = require('../trace');
+var guide = require('../guide');
+
+module.exports = function(env) {
+
+  function DREAM(wpplFn, s, a, options, state, params, step, cont) {
+    this.opts = options;
+
+    if (!_.has(this.opts, 'records')) {
+      throw 'Records of traces and observations are required.';
+    }
+
+    // A list of records, each contains a completed trace
+    // and a list of observed values from a previous execution wpplFn.
+    // Records can be created by using dreamSample.
+    this.records = this.opts.records;
+    this.currRecord;
+
+    this.params = params;
+    this.cont = cont;
+
+    this.wpplFn = wpplFn;
+    this.s = s;
+    this.a = a;
+
+    // The global model includes anything outside of any mapData (level 0) 
+    // The local model includes anything inside of some mapData (level 1+)
+    this.mapDataNestingLevel = 0;
+
+    this.isInsideMapData = function () {
+      return this.mapDataNestingLevel > 0;
+    }
```

Comment:
This works fine, but it's probably best to add this method as a property of `DREAM.prototype` since that's the way all of the other methods (`run`, `sample`, etc.) are added.